src/cascadia/Remoting/Monarch.cpp

```diff
@@ -69,12 +69,31 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
             }
             else
             {
+                uint64_t old = 0;
+
                 // Peasant already had an ID (from an older monarch). Leave that one
                 // be. Make sure that the next peasant's ID is higher than it.
-                _nextPeasantID = providedID >= _nextPeasantID ? providedID + 1 : _nextPeasantID;
+                // If multiple peasants are added concurrently we keep trying to update
+                // until we get to set the new id.
+                while (!_nextPeasantID.compare_exchange_strong(old, providedID + 1))
+                {
+                    old = _nextPeasantID.load();
+                    if (providedID < old)
+                    {
+                        break;
+                    }
+                }

```

Comment:
Strong exchanges aren't needed here as you're looping anways. By using a do-while loop we can get rid of the if condition. This results in:
```cpp
uint64_t current;
do
{
    current = _nextPeasantID.load(std::memory_order_relaxed);
}
while (current <= providedID && !_nextPeasantID.compare_exchange_weak(current, providedID + 1, std::memory_order_relaxed));
```

Reply:
My understanding of `relaxed` is that I could potentially never observe a change from another thread? I admittedly just went with `seq_cst` because it was the lazy option that required the least thinking, and I didn't expect this to occur frequently enough that it would be a huge detriment to performance.

Reply:
`relaxed` means that you don't observe writes to _other_ memory locations, aside from the atomic itself. Even if you use `relaxed` ordering it'll still not permit the compiler to hoist the read out of the loop and not permit the CPU to read stale values of the atomic.

Reply:
I made the change, but it is too late at night for me to reason through it properly right now.